### ISSUE TITLE
Incorrect deserialization of generic collections with numbers

### ISSUE THREAD

**Issue by jkubrynski** [RuedigerMoeller/fast-serialization#81]
I'm serializing ```List<Long>``` to json, which looks as below:
```
{"longList":{"typ":"list","obj":[1,3]}
```
Unfortunately after deserialization list contains ```Integer``` instead of ```Long``` object.

**Comment by RuedigerMoeller** [comment]
Unfortunately this is a json restriction, as there is no way to determine at runtime that "1" should be a long. Because of type erasure its also not easily possible to find the generic type of the collection.
Even worse, there is no general way to figure out, that List is a container and the (hard to obtain) generic type refers to the array elements (fast serialization would be need to read and understand  source code to do that). BUT, it might be possible to hardwire common java collection interfaces such as list, map, but this will require some quirks (especially because of type erasure). I'll keep it as a feature request + thanks for the report.

**Comment by jkubrynski** [comment]
It's in fact not so hard. Of course you'd have to add some checks to use it "in production: :)

```
Field listField = ClassImDeserializing.class.getDeclaredField("myFieldWithGenericCollection")
Type genericType = ((ParameterizedType) listField.getGenericInfo().getGenericType()).getActualTypeArguments()[0];
System.out.println(type.getTypeName()); // prints java.lang.Long
```

**Comment by RuedigerMoeller** [comment]
yeah i am using this in kson parsing also, its just that fst does not know
of the container when parsing the long and the type is not in the
collection but the reference to the collection only ..

Am Mittwoch, 12. August 2015 schrieb Jakub Kubryński :

**Comment by jkubrynski** [comment]
I'm not familiar with FST internals, but it looks ```FSTArrayListSerializer``` when invoking ```readObjectInternal``` could pass info that it expects Long?

**Comment by RuedigerMoeller** [comment]
yep, fst uses referencee type info to optimize. its information present for
free at sender and receiver :). happy to merce a pull btw

Am Mittwoch, 12. August 2015 schrieb Jakub Kubryński :

**Comment by RuedigerMoeller** [comment]
It can't work in general but for simple cases. A serializer gets the reference info (which links to java reflect.field) thrown in. 

```java
readObject(FSTObjectInput in, Object toRead, FSTClazzInfo clzInfo, FSTClazzInfo.FSTFieldInfo referencedBy)
```

However its not always present (e.g. `new Object[] { new List<Long>}`), in addition types can be nested like `Map<String, List<Long>>` ..

An implementation would have to create a new impl for ArrayListSerializer and register it in case of JSon encoding to avoid performance degradation.
